funcx_sdk/funcx/sdk/executor.py

```diff
@@ -161,7 +161,7 @@ def __init__(
         self._tasks_to_send: queue.Queue[
             tuple[FuncXFuture, TaskSubmissionInfo] | tuple[None, None]
         ] = queue.Queue()
-        self._function_registry: dict[t.Callable, str] = {}
+        self._function_registry: dict[t.Callable, str | None] = {}
 
```

Comment:
Ah, why this change?  Are we changing the type of this class variable?  The function should point to it's registration string ... I think without exception.

Reply:
I only made the change to make mypy happy. Below are the errors I was getting.
```
funcx/sdk/executor.py:261: error: Incompatible types in assignment (expression has type "Optional[str]", target has type "str")  [assignment]
funcx/sdk/executor.py:263: error: Incompatible return value type (got "Optional[str]", expected "str")  [return-value]
```

Reply:
Hmm; then I argue that we need to teach mypy what's correct, or understand why mypy is complaining.  because that dictionary, unless I'm missing something, cannot have `None` in it, nor do we want it to.

Reply:
If it's truly broken as of a recent `mypy` or other change, then I encourage making an ancillary PR to address that -- best not to clutter this PR with an unrelated change.

Reply:
The root cause seems to be that the `register_function()` method accepts `function_id` as type str or None ...
https://github.com/funcx-faas/funcX/blob/6e1c1351ad4b9199be468669dce4f23004dc819d/funcx_sdk/funcx/sdk/executor.py#L203-L205

... but then get's added to the `self._function_registry` dict, which only accepts str values.
https://github.com/funcx-faas/funcX/blob/6e1c1351ad4b9199be468669dce4f23004dc819d/funcx_sdk/funcx/sdk/executor.py#L261

The code technically should not allow `function_id` to be None here or when returned, but it's tripping up mypy. To fix without modifying type hints, I've added an `assert function_id is not None` just before updating the registry dict.

Reply:
So, mypy clearly has missed the update to `function_id`: it is guaranteed to be a string at that point.  So, the `assert` is an option.

Another option would be to not reuse the `function_id` variable.  Consider:
```diff
@@ -253,15 +253,14 @@ class FuncXExecutor(concurrent.futures.Executor):
         reg_kwargs = {"function_name": fn.__name__}
         reg_kwargs.update(func_register_kwargs)
         try:
-            function_id = self.funcx_client.register_function(fn, **reg_kwargs)
+            func_id = self.funcx_client.register_function(fn, **reg_kwargs)
         except Exception:
             log.error(f"Unable to register function: {fn.__name__}")
             self.shutdown(wait=False, cancel_futures=True)
             raise
-        assert function_id is not None
-        self._function_registry[fn] = function_id
+        self._function_registry[fn] = func_id
         log.debug("Function registered with id: %s", function_id)
-        return function_id
+        return func_id
 
     def submit(self, fn, *args, **kwargs):
         """
```

Reply:
For posterity, this change has been merged as of PR #991; the plan is to rebase this PR upon `main`.